### DISCUSSION TITLE
PythonTransform cannot return more than 7 items?

### DISCUSSION THREAD

**Discussion by jakeswann1** [bonsai-rx/bonsai#1683] [Q&A] [answered]
Hi all! I'm writing a Bonsai script to control some doors in a radial 8-arm maze, and I'm a bit confused by a behaviour that I've come across. I was wondering if anyone could offer any advice

In brief, I have a PythonTransform node which returns a Tuple with 8 boolean values, each relating to the state of a door (open or closed), which is passed as a Boolean value to an Arduino. 

Items 1 through 7 of the output can be extracted without a problem, but the 8th item is called 'Rest' by default and has the structure 'Tuple[bool]' rather than just 'bool', which is throwing an error. 

Is this an expected behaviour, and what is the recommended fix? I can turn the output into two tuples of four items and pass them to my Arduino that way if need be, but I'd be interested to understand why this happend. I have attached some images to illustrate, as well as the python code within the node, if that helps!

![image](https://github.com/bonsai-rx/bonsai/assets/66915197/3da56610-271e-4478-b0bd-ab255aa58659)
![Screenshot 2024-02-22 171323](https://github.com/bonsai-rx/bonsai/assets/66915197/9545427e-b527-4a2c-b600-e8a4029cd9e2)

```
    from System import Tuple

    # Initialize all doors as closed (True = Closed, False = Open)
    state = Tuple.Create(True, True, True, True, True, True, True, True)
    reward_visited = False

    @returns(Tuple[bool, bool, bool, bool, bool, bool, bool, bool])

    def process(value):
        global state, reward_visited
        dwell_time = value.Item1
        roi = value.Item2
        min_dwell = value.Item3
        delay = value.Item4

        # Detect re-entry into the central octagon and close the last open door if reward_visited is True
        if roi == -1 and reward_visited:
            state = Tuple.Create(True, True, True, True, True, True, True, True)
            reward_visited = False  # Reset for the next cycle
    
        # Handle the initial delay and open all doors
        if roi == -1:  # Assuming -1 indicates waiting in the central octagon
            if dwell_time > delay:
                # Open all doors after the delay
                state = Tuple.Create(False, False, False, False, False, False, False, False)
    
        # Handle animal detected in specific ROIs
        if dwell_time > min_dwell:
            if 1 <= roi <= 8:
                # Close all doors except the one leading to the current arm
                new_state = [True] * 8
                new_state[roi - 1] = False  # Open the door for the current arm
                state = Tuple.Create(*new_state)
            elif 9 <= roi <= 16:
                reward_visited = True
        return state
```

**Comment by bruno-f-cruz** [comment]
Hi,

[Tuples are limited to 7 elements](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/value-tuples). Any reason why you need a Tuple and can't use an array of booleans?

**Reply by jakeswann1** [reply]
Hi Bruno, thanks for your quick reply! I'd happily use an array of Booleans, but I'm a bit confused about how to split the output into items for passing to the DigitalOutput nodes in that case. Could you suggest some syntax that would allow me to parse each item in the array as a separate node?

**Reply by bruno-f-cruz** [reply]
Sure, add your workflow and I will modify it with a possible solution.

**Reply by jakeswann1** [reply]
Thank you! Here it is:
[eight-arm maze control.zip](https://github.com/bonsai-rx/bonsai/files/14386000/eight-arm.maze.control.zip)

**Reply by bruno-f-cruz** [reply, marked as answer]
[eight-arm maze control.zip](https://github.com/bonsai-rx/bonsai/files/14386201/eight-arm.maze.control.zip)

**Reply by jakeswann1** [reply]
Perfect, thank you so much!
